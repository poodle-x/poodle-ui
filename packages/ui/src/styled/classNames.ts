import React from "react";
import { css, CSSObjectSystem, classNames as classNamesCore } from "./";
import { ThemeValue } from "../theme";
import { cssSystem } from "./cssSystem";

export type ClassName =
	| undefined
	| string
	| ((props: any) => string)
	| { [key: string]: boolean };

export interface CSSSystem {
	base?: CSSObjectSystem[];
	check?: boolean;
	overrideClassName?: string | boolean | undefined;
	applies?: (ThemeValue<CSSObjectSystem> | undefined)[];
	override?: ThemeValue<CSSObjectSystem>;
}

export function classNameCompose(data: {
	props: any;
	overrideClassName?: string | boolean | undefined;
	styles: {
		base: CSSObjectSystem[];
		applies?: (ThemeValue<CSSObjectSystem> | undefined)[];
		override?: ThemeValue<CSSObjectSystem>;
	};
}): string {
	const { overrideClassName, styles, props } = data;

	if (typeof overrideClassName === "string") {
		return overrideClassName;
	}

	if (overrideClassName === true) {
		return "";
	}

	return genClassNames({ props, classNames: [styles] });
}

function isCSSSystem(arg: ClassName | CSSSystem): arg is CSSSystem {
	if (typeof arg !== "object") {
		return false;
	}
	return arg.base !== undefined;
}

function genClassNameFromCSSSystem(props: any, sys: CSSSystem) {
	if (typeof sys.overrideClassName === "string") {
		return sys.overrideClassName;
	}

	if (sys.overrideClassName === true || sys.check === false) {
		return "";
	}

	return css(
		cssSystem({
			props,
			...sys,
		})
	);
}

export function genClassNames(data: {
	props: { [key: string]: any };
	classNames: Array<
		ClassName | CSSSystem | ((props: { [key: string]: any }) => CSSSystem)
	>;
}): string {
	const { props, classNames } = data;

	return classNamesCore(
		...classNames
			.filter((s) => {
				return Boolean(s);
			})
			.map((cn) => {
				if (typeof cn === "function") {
					const result = cn(props);

					if (typeof result === "string") {
						return result;
					}

					return genClassNameFromCSSSystem(props, result);
				}

				if (isCSSSystem(cn)) {
					return genClassNameFromCSSSystem(props, cn);
				}

				return cn;
			})
	);
}

export function useClassNames(data: {
	props?: { [key: string]: any };
	theme?: { [key: string]: any };
	lists: {
		[key: string]: {
			props?: { [key: string]: any };
			theme?: { [key: string]: any };
			classNames: Array<ClassName | CSSSystem | ((props: any) => CSSSystem)>;
		};
	};
}): { [key: string]: string } {
	const { props, theme, lists } = data;

	return React.useMemo(() => {
		const result: { [key: string]: string } = {};

		const baseProps = { ...props, theme: theme ? theme : props?.theme };

		Object.keys(lists).forEach((key) => {
			const listKeyData = lists[key];
			const classNames = listKeyData.classNames || [];

			const finalProps = {
				...baseProps,
				theme: listKeyData.theme ? listKeyData.theme : baseProps.theme,
			};
			result[key] = genClassNames({
				props: finalProps,
				classNames,
			});
		});

		return result;
	}, [lists, props, theme]);
}
