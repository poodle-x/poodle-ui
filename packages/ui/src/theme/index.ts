import React from "react";
import {
	ThemeValue,
	SystemThemeConfig,
	BreakpointsThemeConfig,
	SpacingThemeConfig,
	ColorsThemeConfig,
	FontsThemeConfig,
	BordersThemeConfig,
	ShadowThemeConfig,
	LayoutsThemeConfig,
} from "@poodle/system";

import { CSSObjectSystem, Stylesheet } from "../styled";
import { BoxThemeConfig } from "../Box";
import { ButtonThemeConfig } from "../Button";
import { ColumnsThemeConfig } from "../Columns";
import { ColumnThemeConfig } from "../Column";
import { ContainerThemeConfig } from "../Container";
import { TableThemeConfig } from "../Table";
import { TableHeadThemeConfig } from "../TableHead";
import { TableBodyThemeConfig } from "../TableBody";
import { TableRowThemeConfig } from "../TableRow";
import { TableCellThemeConfig } from "../TableCell";
import { VisuallyHiddenThemeConfig } from "../VisuallyHidden";
import { CheckControlThemeConfig } from "../CheckControl";
import { CheckboxThemeConfig } from "../Checkbox";
import { RadioThemeConfig } from "../Radio";
import { InputThemeConfig } from "../Input";
import { InputAdornmentThemeConfig } from "../InputAdornment";
import { IconButtonThemeConfig } from "../IconButton";
import { ModalThemeConfig } from "../Modal";
import { ModalHeaderThemeConfig } from "../ModalHeader";
import { ModalBodyThemeConfig } from "../ModalBody";
import { ModalFooterThemeConfig } from "../ModalFooter";
import { SelectThemeConfig } from "../Select";
import { TabsThemeConfig } from "../Tabs";
import { TabThemeConfig } from "../Tab";
import { TabListThemeConfig } from "../TabList";
import { TabPanelThemeConfig } from "../TabPanel";

export { Stylesheet };

export { useTheme } from "./useTheme";

export type DirectionThemeConfig = "rtl" | "ltr";

export interface GlobalThemeConfig {
	styles?: ThemeValue<CSSObjectSystem>;
}

export type IconType = React.FC<
	React.PropsWithoutRef<React.SVGProps<SVGSVGElement>> &
		React.RefAttributes<SVGSVGElement>
>;

export interface BaseIcons {
	RadioButton?: IconType;
	RadioCheckedButton?: IconType;
	CheckboxButton?: IconType;
	CheckboxCheckedButton?: IconType;
	CheckboxIndeterminateButton?: IconType;
	ArrowDown?: IconType;
}

export interface ThemeConfig extends SystemThemeConfig {
	global?: GlobalThemeConfig;
	direction?: ThemeValue<DirectionThemeConfig>;
	baseIcons?: BaseIcons;
	Box?: BoxThemeConfig;
	Button?: ButtonThemeConfig;
	Columns?: ColumnsThemeConfig;
	Column?: ColumnThemeConfig;
	Container?: ContainerThemeConfig;
	Table?: TableThemeConfig;
	TableHead?: TableHeadThemeConfig;
	TableBody?: TableBodyThemeConfig;
	TableRow?: TableRowThemeConfig;
	TableCell?: TableCellThemeConfig;
	VisuallyHidden?: VisuallyHiddenThemeConfig;
	CheckControl?: CheckControlThemeConfig;
	Checkbox?: CheckboxThemeConfig;
	Radio?: RadioThemeConfig;
	Input?: InputThemeConfig;
	InputAdornment?: InputAdornmentThemeConfig;
	IconButton?: IconButtonThemeConfig;
	Modal?: ModalThemeConfig;
	ModalHeader?: ModalHeaderThemeConfig;
	ModalBody?: ModalBodyThemeConfig;
	ModalFooter?: ModalFooterThemeConfig;
	Select?: SelectThemeConfig;
	Tabs?: TabsThemeConfig;
	Tab?: TabThemeConfig;
	TabList?: TabListThemeConfig;
	TabPanel?: TabPanelThemeConfig;
}

export const defaultGlobal = (
	overrideTheme: ThemeConfig
): GlobalThemeConfig => {
	return {
		styles: overrideTheme.global?.styles,
	};
};

export const defaultBreakpoints = (
	overrideTheme: ThemeConfig
): BreakpointsThemeConfig => {
	return {
		scale: ["sm", "md", "lg", "xl"],
		...overrideTheme.breakpoints,
		sets: {
			sm: 576,
			md: 768,
			lg: 992,
			xl: 1200,
			xxl: 1400,
			...overrideTheme.breakpoints?.sets,
		},
	};
};

export const defaultSpacing = (
	overrideTheme: ThemeConfig
): SpacingThemeConfig => {
	return {
		scale: 4,
		...overrideTheme.spacing,
		sets: {
			s: 4,
			m: 8,
			l: 16,
			...overrideTheme.spacing?.sets,
		},
	};
};

export const defaultFonts = (overrideTheme: ThemeConfig): FontsThemeConfig => {
	return {
		scale: "1rem",
		...overrideTheme.fonts,
		base: {
			fontFamily:
				'-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"',
			fontStyle: "normal",
			fontWeight: "normal",
			lineHeight: "125%",
			fontSize: "1rem",
			...overrideTheme.fonts?.base,
		},
		weights: {
			normal: 400,
			semibold: 600,
			bold: 700,
			...overrideTheme.fonts?.weights,
		},
		sets: {
			input: {
				fontSize: "1rem",
				fontWeight: "normal",
				lineHeight: "125%",
			},
			button: {
				fontSize: "1rem",
				fontWeight: "bold",
				lineHeight: "150%",
			},
			label: {
				fontStyle: "normal",
				fontWeight: "bold",
				fontSize: "0.875rem",
				lineHeight: "110%",
			},
			helper: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "0.875rem",
				lineHeight: "110%",
			},
			checkControlLabelMedium: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "1rem",
				lineHeight: "110%",
			},
			checkControlLabelSmall: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "0.875rem",
				lineHeight: "110%",
			},
			checkControlLabelLarge: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "1.125rem",
				lineHeight: "110%",
			},
			modalTitle: {
				fontStyle: "normal",
				fontWeight: "bold",
				fontSize: "1.5rem",
				lineHeight: "133.33%",
			},
			...overrideTheme.fonts?.sets,
		},
	};
};

export const defaultColor = (overrideTheme: ThemeConfig): ColorsThemeConfig => {
	return {
		modes: {
			...overrideTheme.colors?.modes,
		},
		sets: {
			text: "#282828",
			disabledText: "#A6A6A6",

			primary: {
				light: "#D1EBFF",
				base: "#00478A",
				dark: "#003A70",
				textOnLight: "#00478A",
				textOnBase: "#fff",
				textOnDark: "#fff",
			},

			negative: {
				base: "#AF1308",
				dark: "#8D1007",
				light: "#FCF3F3",
				textOnLight: "#AF1308",
				textOnBase: "#fff",
				textOnDark: "#fff",
			},

			positive: {
				base: "#006122",
				dark: "#004D1B",
				light: "#E5FFEE",
				textOnLight: "#006122",
				textOnBase: "#fff",
				textOnDark: "#fff",
			},

			warn: {
				base: "#FFBB33",
				dark: "#FAAB00",
				light: "#FFF1DB",
				textOnLight: "#282828",
				textOnBase: "#282828",
				textOnDark: "#282828",
			},

			bg: {
				base: "#F8F8F8",
				100: "#E0E0E0",
				textOn100: "#282828",
				200: "#CCCCCC",
				textOn200: "#282828",
			},

			modal: {
				base: "#F8F8F8",
				footer: "#F0F0F0",
			},

			border: {
				base: "#CCCCCC",
			},

			...overrideTheme.colors?.sets,
		},
	};
};

export const defaultBorder = (
	overrideTheme: ThemeConfig
): BordersThemeConfig => {
	return {
		sets: {
			base: {
				style: "solid",
				width: "1px",
				color: "border",
			},
			...overrideTheme?.borders?.sets,
		},
		radius: {
			base: "4px",
			...overrideTheme?.borders?.radius,
		},
	};
};

export const defaultBoxShadow = (
	overrideTheme: ThemeConfig
): ShadowThemeConfig => {
	return {
		boxShadowSets: {
			lv1: "1px 2px 2px -1px rgba(0,0,0,0.4)",
			lv2: "2px 2px 5px 0px rgba(0,0,0,0.4)",
			lv3: "4px 4px 5px 0px rgba(0,0,0,0.4)",
			...overrideTheme?.shadows?.boxShadowSets,
		},
	};
};

export const defaultLayouts = (
	overrideTheme: ThemeConfig
): LayoutsThemeConfig => {
	return {
		zIndices: {
			modal: 2000,
			...overrideTheme?.layouts?.zIndices,
		},
	};
};

export function createTheme(overrideTheme: ThemeConfig = {}): ThemeConfig {
	return {
		direction: "ltr",
		...overrideTheme,
		global: defaultGlobal(overrideTheme),
		spacing: defaultSpacing(overrideTheme),
		breakpoints: defaultBreakpoints(overrideTheme),
		colors: defaultColor(overrideTheme),
		fonts: defaultFonts(overrideTheme),
		borders: defaultBorder(overrideTheme),
		shadows: defaultBoxShadow(overrideTheme),
		layouts: defaultLayouts(overrideTheme),
	};
}

export type StylesConfig<T extends string> = {
	[K in T]: ThemeValue<CSSObjectSystem>;
};

export interface StandardComponentProps {
	variant?: string;
	themeExtend?: ThemeConfig;
	theme?: ThemeConfig;
	sizeStyle?: string;
	colorStyle?: string;
}

export interface StandardThemeConfig<
	T extends string,
	Props extends { [key: string]: any }
> {
	defaultProps?: Partial<Props>;
	styles?: Partial<StylesConfig<T>>;
	variants?: {
		[key: string]: Partial<StylesConfig<T>>;
	};
	sizes?: {
		[key: string]: Partial<StylesConfig<T>>;
	};
	colorStyles?: {
		[key: string]: Partial<StylesConfig<T>>;
	};
	modes?: {
		[key: string]: {
			styles?: Partial<StylesConfig<T>>;
			variants?: {
				[key: string]: Partial<StylesConfig<T>>;
			};
			sizes?: {
				[key: string]: Partial<StylesConfig<T>>;
			};
			colorStyles?: {
				[key: string]: Partial<StylesConfig<T>>;
			};
		};
	};
	overrides?: Partial<StylesConfig<T>>;
	overrideClasses?: Partial<
		{
			[K in T]: boolean | string;
		}
	>;
}

export * from "./system";
