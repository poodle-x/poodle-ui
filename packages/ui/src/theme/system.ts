import {
	colorsSystem,
	fontsSystem,
	layoutSystem,
	spacingSystem,
	compose,
	createSystem,
	getMode,
	FontsProps,
	ColorsProps,
	SpacingProps,
	LayoutProps,
	BordersProps,
	bordersSystem,
	flexSystem,
	FlexProps,
	pseudosSystem,
	PseudosProps,
	shadowSystem,
	ShadowProps,
	GridProps,
	gridSystem,
	ListProps,
	listSystem,
	SystemConfigList,
	ResponsiveProp,
} from "@poodle/system";
import { CSSSystem } from "../styled";
import { StandardThemeConfig } from "./index";
import * as CSS from "csstype";

const additionalSystem: SystemConfigList = {
	objectFit: true,
	objectPosition: true,
	transition: true,
	cursor: true,
	pointerEvents: true,
};

export interface AdditionalSystemProps {
	/**
	 * The CSS `object-fit` property
	 */
	objectFit?: ResponsiveProp<CSS.Property.ObjectFit>;
	/**
	 * The CSS `object-fit` property
	 */
	objectPosition?: ResponsiveProp<CSS.Property.ObjectPosition>;
	/**
	 * The CSS `transition` property
	 */
	transition?: ResponsiveProp<CSS.Property.Transition>;
	/**
	 * The CSS `cursor` property
	 */
	cursor?: ResponsiveProp<CSS.Property.Cursor>;
	/**
	 * The CSS `cursor` property
	 */
	pointerEvents?: ResponsiveProp<CSS.Property.PointerEvents>;
}

const allSystem = compose(
	colorsSystem,
	fontsSystem,
	layoutSystem,
	spacingSystem,
	bordersSystem,
	flexSystem,
	pseudosSystem,
	shadowSystem,
	gridSystem,
	listSystem,
	additionalSystem
);

export const system = createSystem(allSystem);

interface CssSystemProps
	extends SpacingProps,
		LayoutProps,
		FontsProps,
		ColorsProps,
		BordersProps,
		ShadowProps,
		FlexProps,
		GridProps,
		ListProps {}

export interface SystemProps
	extends SpacingProps,
		LayoutProps,
		FontsProps,
		ColorsProps,
		BordersProps,
		ShadowProps,
		FlexProps,
		PseudosProps<CssSystemProps>,
		GridProps,
		ListProps,
		AdditionalSystemProps {}

const additionalProps = [
	"textStyle",
	"variant",
	"themeExtend",
	"theme",
	"sizeStyle",
	"colorStyle",
];

export function removeSystemProps(props: { [key: string]: any }) {
	let resultProps = {};
	Object.keys(props).forEach((prop) => {
		let found = false;
		for (const name in allSystem) {
			if (prop === name || additionalProps.indexOf(prop) >= 0) {
				found = true;
				break;
			}
		}

		if (!found && ["sx"].indexOf(prop) === -1) {
			resultProps = {
				...resultProps,
				[prop]: props[prop],
			};
		}
	});

	return resultProps;
}

export function createCSSSystemStandard<Key extends string>(data: {
	key: Key;
	config?: StandardThemeConfig<Key, any>;
	props: { [key: string]: any };
	base?: CSSSystem["base"];
}): CSSSystem {
	const { key, props, config, base } = data;

	const { variant = "", sizeStyle = "", colorStyle = "" } = props;

	const mode = getMode(props);

	const overrideClassName = config?.overrideClasses?.[key];

	return {
		base,
		overrideClassName,
		applies: [
			config?.styles?.[key],
			config?.modes?.[mode]?.styles?.[key],
			config?.variants?.[variant]?.[key],
			config?.modes?.[mode]?.variants?.[variant]?.[key],
			config?.sizes?.[sizeStyle]?.[key],
			config?.modes?.[mode]?.sizes?.[sizeStyle]?.[key],
			config?.colorStyles?.[colorStyle]?.[key],
			config?.modes?.[mode]?.colorStyles?.[colorStyle]?.[key],
		],
		override: config?.overrides?.[key],
	};
}

export * from "@poodle/system";
