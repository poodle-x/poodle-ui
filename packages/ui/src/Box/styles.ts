import { system, createCSSSystemStandard } from "../theme";
import { BoxProps } from "./Box";
import { CSSSystem, css } from "../styled";

export const Root = (props: BoxProps): CSSSystem => {
	const { theme } = props;
	return createCSSSystemStandard({
		key: "root",
		config: theme?.Box,
		props,
		base: [
			{
				boxSizing: "border-box",
			},
		],
	});
};

export const SX = (props: BoxProps) => {
	const { sx } = props;

	if (!sx) return "";

	return css(system({ ...sx, theme: props.theme }));
};

export const Box = (props: BoxProps) => {
	const { children, ...otherProps } = props;

	return css(system(otherProps, { returnOnlySystemProps: true }));
};
