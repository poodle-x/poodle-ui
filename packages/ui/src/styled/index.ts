import { css as cssCore, cx as cxCore } from "@emotion/css";
import classNames from "classnames";
import { ClassNamesArg, Interpolation } from "@emotion/react";
import * as CSS from "csstype";
import { SystemProps } from "../theme";
import { cssSystem } from "./cssSystem";

function css(...args: Array<Interpolation<any>>) {
	return cssCore(...(args as any));
}

function cx(...classNames: Array<ClassNamesArg>) {
	return cxCore(...classNames);
}

export { cssSystem };

export { css, cx, classNames };

export * from "./classNames";

export type Stylesheet<MP = undefined> = Interpolation<MP>;

export type CSSProperties = CSS.PropertiesFallback<number | string>;

export type CSSPropertiesWithMultiValues = {
	[K in keyof CSSProperties]:
		| CSSProperties[K]
		| Array<Extract<CSSProperties[K], string>>;
};

export type CSSPropertiesWithSystem =
	| CSSPropertiesWithMultiValues
	| SystemProps;

export type CSSPseudoPropertiesWithSystem = {
	[K in CSS.Pseudos]?: AllCSSPropertiesWithSystem;
};

export type AllCSSPropertiesWithSystem = CSSPropertiesWithSystem &
	CSSPseudoPropertiesWithSystem;

export type CSSObjectSystem =
	| AllCSSPropertiesWithSystem
	| { [key: string]: CSSObjectSystem };
