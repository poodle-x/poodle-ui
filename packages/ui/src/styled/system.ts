import { ClassNamesArg } from "@emotion/react";
import { BoxProps } from "../Box";
import { getThemeValue, system, ThemeValue } from "../theme";
import { CSSObjectSystem, CSSProperties, cx, DashSX } from ".";

export function joinDashSX(...agr: (DashSX | undefined)[]): DashSX[] {
	const result: DashSX[] = [];
	agr.forEach((d) => {
		if (d) {
			result.push(d);
		}
	});
	return result;
}

export interface CSSSystem {
	base?: CSSObjectSystem[];
	check?: boolean;
	applies?: (ThemeValue<CSSObjectSystem> | undefined)[];
	override?: ThemeValue<CSSObjectSystem>;
}

export function runCSSSystem(data: {
	props: { [key: string]: any };
	cssSystem: CSSSystem;
}):
	| Array<{ [p: string]: any }>
	| (Array<{ [p: string]: any }> | CSSProperties | CSSProperties[])[] {
	const { props, cssSystem } = data;
	const { applies = [], override, base = [] } = cssSystem;

	const overrideTheme = getThemeValue<CSSObjectSystem | undefined>(
		props,
		override,
		undefined
	);

	if (overrideTheme) {
		return system(
			{ ...overrideTheme, theme: props.theme },
			{
				sendProps: props,
			}
		);
	}

	return [
		...base.map((b) => {
			return system(
				{ ...b, theme: props.theme },
				{
					sendProps: props,
				}
			);
		}),
		...applies.reduce<(CSSObjectSystem | CSSObjectSystem[])[]>((prev, s) => {
			const v = getThemeValue(props, s, undefined);

			if (v) {
				return [
					...prev,
					system({ ...v, theme: props.theme }, { sendProps: props }),
				];
			}

			return prev;
		}, []),
	];
}

export function getCSSSystemBoxProps<
	ComponentProps = BoxProps,
	PartProps = BoxProps
>(opts: {
	isLocalTheme?: boolean;
	isRoot?: boolean;
	componentProps?: ComponentProps;
	partProps?: PartProps;
	fnCSSSystem:
		| ((props: ComponentProps) => CSSSystem)
		| ((props: ComponentProps) => CSSSystem)[];
	baseClassName?: ClassNamesArg[];
}) {
	const {
		isLocalTheme,
		isRoot,
		componentProps = {},
		fnCSSSystem,
		baseClassName,
		partProps = {},
	} = opts;

	const runReturned: DashSX[] = [];
	if (Array.isArray(fnCSSSystem)) {
		fnCSSSystem.forEach((f) => {
			runReturned.push(
				runCSSSystem({
					props: componentProps,
					cssSystem: f(componentProps as any),
				})
			);
		});
	} else {
		runReturned.push(
			runCSSSystem({
				props: componentProps,
				cssSystem: fnCSSSystem(componentProps as any),
			})
		);
	}
	const className: (ClassNamesArg[] | any)[] = [baseClassName];

	if (isRoot) {
		className.push((componentProps as BoxProps).className);
		if ((componentProps as BoxProps)._sx) {
			runReturned.push((componentProps as any)._sx);
		}
	}

	if ((partProps as BoxProps)._sx) {
		runReturned.push((partProps as any)._sx);
	}

	className.push((partProps as BoxProps).className);

	const componentTheme = isLocalTheme
		? (componentProps as BoxProps).theme
		: undefined;

	return {
		...partProps,
		theme: (partProps as BoxProps).theme
			? (partProps as BoxProps).theme
			: componentTheme,
		className: cx(className),
		_sx: runReturned,
	};
}

export function cssSystem(data: {
	props: { [key: string]: any };
	base?: CSSObjectSystem[];
	applies?: (ThemeValue<CSSObjectSystem> | undefined)[];
	override?: ThemeValue<CSSObjectSystem>;
}):
	| Array<{ [p: string]: any }>
	| (Array<{ [p: string]: any }> | CSSProperties | CSSProperties[])[] {
	const { props, applies = [], override, base = [] } = data;

	const overrideTheme = getThemeValue<CSSObjectSystem | undefined>(
		props,
		override,
		undefined
	);

	if (overrideTheme) {
		return system(
			{ ...overrideTheme, theme: props.theme },
			{
				sendProps: props,
			}
		);
	}

	return [
		...base.map((b) => {
			return system(
				{ ...b, theme: props.theme },
				{
					sendProps: props,
				}
			);
		}),
		...applies.reduce<(CSSObjectSystem | CSSObjectSystem[])[]>((prev, s) => {
			const v = getThemeValue(props, s, undefined);

			if (v) {
				return [
					...prev,
					system({ ...v, theme: props.theme }, { sendProps: props }),
				];
			}

			return prev;
		}, []),
	];
}
