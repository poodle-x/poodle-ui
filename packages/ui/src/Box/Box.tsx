import React from "react";
import useDefaultProps from "../utils/useDefaultProps";
import * as styles from "./styles";
import {
	removeSystemProps,
	StandardComponentProps,
	SystemProps,
} from "../theme";
import { useClassNames } from "../styled";
import { CSSObjectSystem } from "../styled";

type As<P = any> = React.ElementType<P> | string;

type PropsWithAs<P, T extends As<P>> = P &
	Omit<
		React.ComponentProps<T extends string ? React.ElementType<T> : T>,
		"as" | keyof P
	> & {
		as?: T;
		children?: React.ReactNode;
	};

type Component<T extends As, O> = {
	<TT extends As>(props: PropsWithAs<O, TT> & { as: TT }): JSX.Element;
	(props: PropsWithAs<O, T>): JSX.Element;
	displayName?: string;
};

export interface LocalBoxProps {
	/**
	 * Class apply to root element.
	 */
	className?: string;

	children?: React.ReactNode;
	/**
	 * Change the element will be rendered.
	 */
	as?: string | React.ComponentType<any>;
}

export interface BoxProps
	extends Omit<React.AllHTMLAttributes<HTMLElement>, keyof SystemProps | "as">,
		React.RefAttributes<HTMLElement>,
		SystemProps,
		StandardComponentProps,
		LocalBoxProps {
	sx?: CSSObjectSystem;
	[key: string]: any;
}

export const Box = React.forwardRef<HTMLElement, BoxProps>((_props, ref) => {
	const props = useDefaultProps<BoxProps>(_props, {
		themeDefaultProps: (t) => {
			return t?.Box?.defaultProps;
		},
	});

	const {
		themeExtend,
		theme: themeOverride,
		className = "",
		children,
		as = "div",
		...otherProps
	} = props;

	const htmlProps = removeSystemProps(otherProps);

	const classes = useClassNames({
		props,
		lists: React.useMemo(() => {
			return {
				root: {
					classNames: [
						"poodle-box",
						styles.Root,
						className,
						styles.SX,
						styles.Box,
					],
				},
			};
		}, [className]),
	});

	return React.createElement(
		as,
		{ ref, className: classes.root, ...htmlProps },
		children
	);
}) as Component<As, BoxProps>;

Box.displayName = "PoodleBox";
