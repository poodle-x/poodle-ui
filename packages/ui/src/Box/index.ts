import { Box, BoxProps } from "./Box";
import { StandardThemeConfig } from "../theme";
import * as styles from "./styles";

export { styles as boxStyles };

export * from "./Box";

export default Box;

export type BoxStyleKeys = "root";

export interface BoxThemeConfig
	extends StandardThemeConfig<BoxStyleKeys, BoxProps> {}
