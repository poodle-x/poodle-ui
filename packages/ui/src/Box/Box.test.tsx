import React from "react";
import { Stylesheet, css, useClassNames } from "../styled";
import { ThemeConfig } from "../theme";
import { render } from "../utils/test";
import { Box, BoxProps } from "./Box";

describe("<Box />", () => {
	it("should render correctly", () => {
		const { container, rerender } = render(<Box />);

		expect(container.firstChild).toMatchSnapshot();

		rerender(<Box>test</Box>);

		expect(container.firstChild).toMatchSnapshot();
	});

	it("`ref` and `as` props should work", () => {
		const mockRef: React.RefObject<HTMLParagraphElement> = { current: null };

		render(<Box as="p" ref={mockRef} />);

		expect(mockRef.current).not.toBeNull();

		expect((mockRef.current as HTMLParagraphElement).tagName).toEqual("P");
	});

	it("should render correctly with aria* props", () => {
		const { container } = render(<Box aria-label="label" aria-live="polite" />);
		expect(container.firstChild).toMatchSnapshot();
	});

	it("box styles should always on top", () => {
		const styles = () => {
			const style: Stylesheet = {
				padding: 20,
			};
			return css(style);
		};

		const Hello: React.FC<BoxProps> = (props) => {
			const { children } = props;

			const classes = useClassNames({
				props,
				lists: {
					root: {
						classNames: [styles],
					},
				},
			});

			return (
				<Box {...props} className={classes.root}>
					{children}
				</Box>
			);
		};

		const { container } = render(<Hello p={50}>test</Hello>);

		expect(container.firstChild).toMatchSnapshot();
	});

	describe("theming", () => {
		const themeTest: ThemeConfig = {
			mode: "mode",
			Box: {
				modes: {
					mode: {
						styles: {
							root: {
								letterSpacing: 250,
							},
						},
					},
				},
				styles: {
					root: {
						top: 0,
						textIndent: 1,
					},
				},
			},
		};

		it("should extend theme in global", () => {
			const { container } = render(
				<Box m={20} p={20}>
					Button
				</Box>,
				{
					theme: themeTest,
				}
			);
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should extend theme in local", () => {
			const { container } = render(
				<Box m={20} p={20} themeExtend={themeTest}>
					Button
				</Box>,
				{
					theme: themeTest,
				}
			);
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should override styles in global", () => {
			const { container } = render(
				<Box m={20} p={20} themeExtend={themeTest}>
					Button
				</Box>,
				{
					theme: {
						Box: {
							overrides: {
								root: {
									margin: 1,
								},
							},
						},
					},
				}
			);
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should override styles in local", () => {
			const { container } = render(
				<Box
					m={20}
					p={20}
					themeExtend={{
						Box: {
							overrides: {
								root: {
									margin: 2,
								},
							},
						},
					}}
				>
					Button
				</Box>,
				{
					theme: {
						Box: {
							overrides: {
								root: {
									margin: 1,
								},
							},
						},
					},
				}
			);
			expect(container.firstChild).toMatchSnapshot();
		});
	});
});
