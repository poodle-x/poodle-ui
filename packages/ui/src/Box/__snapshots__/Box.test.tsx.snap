// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Box /> box styles should always on top 1`] = `
.c0 {
  box-sizing: border-box;
}

.c1 {
  padding: 20px;
}

.c2 {
  padding: 50px;
}

<div
  class="poodle-box c0 c1 c2"
>
  test
</div>
`;

exports[`<Box /> should render correctly 1`] = `
.c0 {
  box-sizing: border-box;
}

<div
  class="poodle-box c0 c1"
/>
`;

exports[`<Box /> should render correctly 2`] = `
.c0 {
  box-sizing: border-box;
}

<div
  class="poodle-box c0 c1"
>
  test
</div>
`;

exports[`<Box /> should render correctly with aria* props 1`] = `
.c0 {
  box-sizing: border-box;
}

<div
  aria-label="label"
  aria-live="polite"
  class="poodle-box c0 c1"
/>
`;

exports[`<Box /> theming should extend theme in global 1`] = `
.c0 {
  box-sizing: border-box;
  top: 0;
  text-indent: 1px;
  letter-spacing: 250px;
}

.c1 {
  margin: 20px;
  padding: 20px;
}

<div
  class="poodle-box c0 c1"
>
  Button
</div>
`;

exports[`<Box /> theming should extend theme in local 1`] = `
.c0 {
  box-sizing: border-box;
  top: 0;
  text-indent: 1px;
  letter-spacing: 250px;
}

.c1 {
  margin: 20px;
  padding: 20px;
}

<div
  class="poodle-box c0 c1"
>
  Button
</div>
`;

exports[`<Box /> theming should override styles in global 1`] = `
.c0 {
  margin: 1px;
}

.c1 {
  margin: 20px;
  padding: 20px;
}

<div
  class="poodle-box c0 c1"
>
  Button
</div>
`;

exports[`<Box /> theming should override styles in local 1`] = `
.c0 {
  margin: 2px;
}

.c1 {
  margin: 20px;
  padding: 20px;
}

<div
  class="poodle-box c0 c1"
>
  Button
</div>
`;
